docs:  rollout concept doc (#819)
Azure/fleet: docs/concepts/SafeRollout/README.md

cluster has label `env: prod`. The hub cluster has a namespace called `test-ns` with a deployment in it.

The `ClusterResourcePlacement` spec is defined as follows:

```yaml
spec:
  resourceSelectors:
    - group: ""
      kind: Namespace
      version: v1
      name: test-ns
  policy:
    placementType: PickN
    numberOfClusters: 3
    affinity:
      clusterAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          clusterSelectorTerms:
            - labelSelector:
                matchLabels:
                  env: prod
  strategy:
    rollingUpdate:
      maxUnavailable: 1
      maxSurge: 1
```

The rollout will be as follows:

- We try to pick 3 clusters out of 4, for this scenario let's say we pick cluster-1, cluster-2 & cluster-3.
- Since we can't track the initial availability for the deployment, we rollout the namespace with deployment to 
cluster-1, cluster-2 & cluster-3
- Then we update the deployment with a bad image name to update the resource in place on cluster-1, cluster-2 & cluster-3
- But since we have `maxUnavailable` set to 1, we will rollout the bad image name update for deployment to one of the clusters 

> (which cluster the resource is rolled out to first is non-deterministic)

Reviewer: ```suggestion
(which cluster the resource is rolled out to first is non-deterministic).

```

- Since we can't track the initial availability for the deployment, we rollout the namespace with deployment to cluster-1
and cluster-2 and wait till they become available.

Then we update the `ClusterResourcePlacement` spec to the following:

```yaml
spec:
  resourceSelectors:
    - group: ""
      kind: Namespace
      version: v1          
      name: test-ns
  policy:
    placementType: PickN
    numberOfClusters: 2
    affinity:
      clusterAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          clusterSelectorTerms:
              - labelSelector:
                  matchLabels:
                    loc: east
  strategy:
    rollingUpdate:
      maxSurge: 2
```

The rollout will be as follows:

- We try to pick clusters (cluster-3 and cluster-4) by specifying the label selector `loc: east`.
- But this time around since we have `maxSurge` set to 2 we are saying we can propagate resources to a maximum of 
4 clusters but our target number of clusters specified is 2, we will rollout the namespace with deployment to both 
cluster-3 and cluster-4 before removing the deployment from cluster-1 and cluster-2. 

> - And since `maxUnavailable` is always set to 1 by default, we will try to remove the resource from cluster-1 first and 

Author: ```suggestion
- And since `maxUnavailable` is set to 25% by default, we remove the resource from one of the existing clusters

```

as available in the target clusters when they meet the criteria we defined.

### How It Works
We have an agent running in the target cluster to check the status of the resources. We have specific criteria for each 
of the following resources to determine if they are available or not. Here are the list of resources we support:

#### Deployment
We only mark a `Deployment` as available when all its pods are running, ready and updated according to the latest spec. 

#### DaemonSet 
We only mark a `DaemonSet` as available when all its pods are available and updated according to the latest spec on all 
desired scheduled nodes.

#### StatefulSet
We only mark a `StatefulSet` as available when all its pods are running, ready and updated according to the latest revision.

#### Job
We only mark a `Job` as available when it has at least one succeeded pod or one ready pod.

#### Service
For `Service` based on the service type the availability is determined as follows:

- For `ClusterIP` & `NodePort` service, we mark it as available when a cluster IP is assigned.
- For `LoadBalancer` service, we mark it as available when a `LoadBalancerIngress` has been assigned along with an IP or Hostname.

> - For `ExternalName` service, we don't know how to determine the availability.

Reviewer: ```suggestion
- For `ExternalName` service, checking availability is not supported, so it will be marked as available with not trackable reason.

```
